sklearn/utils/tests/test_extmath.py

```diff
@@ -453,6 +455,116 @@ def naive_log_logistic(x):
     assert_array_almost_equal(log_logistic(extreme_x), [-100, 0])
 
 
+def test_incremental_weighted_mean_and_variance_simple():
+    rng = np.random.RandomState(42)
+    mult = 10
+    X = rng.rand(1000, 20)*mult
+    sample_weight = rng.rand(X.shape[0]) * mult
+    mean, var, _ = _incremental_weighted_mean_and_var(X, sample_weight,
+                                                      0, 0, 0)
+
+    mean_exp = np.average(X, weights=sample_weight, axis=0)
+    var_exp = np.average(X**2, weights=sample_weight, axis=0) - mean_exp**2
+    assert_almost_equal(mean, mean_exp)
+    assert_almost_equal(var, var_exp)
+
+
+def test_incremental_weighted_mean_and_variance():
+
+    # Testing of correctness and numerical stability
+    def test(X, sample_weight, mean_exp=None, var_exp=None):
+        n = X.shape[0]
+        if mean_exp is None:
+            mean_exp = \
+                _safe_accumulator_op(
+                    np.average, X, weights=sample_weight, axis=0)
+        if var_exp is None:
+            var_exp = \
+                _safe_accumulator_op(
+                    np.average, (X-mean_exp)**2, weights=sample_weight, axis=0)
+        for chunk_size in [1, n//10 + 1, n//4 + 1, n//2 + 1, n]:
+            last_mean, last_weight_sum, last_var = 0, 0, 0
+            for batch in gen_batches(n, chunk_size):
+                last_mean, last_var, last_weight_sum = \
+                    _incremental_weighted_mean_and_var(X[batch],
+                                                       sample_weight[batch],
+                                                       last_mean,
+                                                       last_var,
+                                                       last_weight_sum)
+            assert_allclose(last_mean, mean_exp)
+            assert_allclose(last_var, var_exp, atol=1e-6)
+
+    HIGH_MEAN = 1e7
+    LOW_VAR = 1e-8
+    HIGH_VAR = 1e5
+    NORMAL_MEAN = 0.0
+    NORMAL_VAR = 1.0
+    SIZE = (100, 20)
+
+    rng = np.random.RandomState(42)
+    NORMAL_WEIGHT = \
+        rng.normal(loc=NORMAL_MEAN, scale=NORMAL_VAR, size=(SIZE[0],))
+    ALMOST_ZERO_WEIGHT = \
+        rng.normal(loc=NORMAL_MEAN, scale=LOW_VAR, size=(SIZE[0],))
+    ALMOST_ONES_WEIGHT = rng.normal(loc=1.0, scale=LOW_VAR, size=(SIZE[0],))
+    JUST_WEIGHT = rng.normal(loc=10.0, scale=NORMAL_VAR, size=(SIZE[0],))
+    HIGH_WEIGHT = rng.normal(loc=HIGH_MEAN, scale=NORMAL_VAR, size=(SIZE[0],))
+    ONES_WEIGHT = np.ones(SIZE[0])
+
+    means = [NORMAL_MEAN, HIGH_MEAN, -HIGH_MEAN]
+    vars = [NORMAL_VAR, LOW_VAR, HIGH_VAR]
+    weights = [NORMAL_WEIGHT, ALMOST_ONES_WEIGHT,
+               JUST_WEIGHT, ALMOST_ZERO_WEIGHT, HIGH_WEIGHT]
+    means_vars = [(m, v) for m in means for v in vars]
+
+    # Comparing with weighted np.average
+    for mean, var in means_vars:
+        X = rng.normal(loc=mean, scale=var, size=SIZE)
+        for i, weight in enumerate(weights):
+            test(X, weight)
```

Comment:
Ideally we should refactor this to use `pytest.mark.parametrize`

Reply:
I don't know how to use pytest.mark.parametrize inside a method and I'm not sure if parametrizing the most 'outside' method is good idea. I'm open for any suggestions.

Reply:
@panpiort8 Something like
```python
@pytest.mark.parametrize("mean", [0, 1e7, -1e7])
@pytest.mark.parametrize("var", [1, 1e-8, 1e5])
def test_incremental_weighted_mean_and_variance(mean, var):
...
```
Then you just use `mean` and `var` inside your test function and save some code, i.e. the loops.